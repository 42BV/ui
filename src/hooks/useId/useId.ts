import { useEffect, useState } from 'react';
import { uniqueId } from 'lodash';

type UseIdConfig = {
  /**
   * Optionally the id to use. When no id is provided a unique id
   * is generated.
   */
  id?: string;
};

/**
 * Either returns the id provided or when it is not defined generates
 * a unique stable id and returns that id.
 *
 * @param {UseIdConfig} useIdConfig The configuration for useId
 * @returns {string} The id.
 */
export function useId({ id }: UseIdConfig): string {
  const [ innerId, setInnerId ] = useState(() => id ?? uniqueId());
  useEffect(() => {
    if (id) {
      setInnerId(id);
    }
  }, [ id ]);

  return innerId;
}
