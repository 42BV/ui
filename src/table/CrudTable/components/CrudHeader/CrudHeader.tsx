import { EpicCellLayout } from '../../../EpicTable/cells/EpicCellLayout/EpicCellLayout';
import { EpicSort } from '../../../EpicTable/widgets/EpicSort/EpicSort';
import { Input } from '../../../../form/Input/Input';
import { EpicTableSortDirection } from '../../../EpicTable/types';
import { LabelForOption } from '../../../../form/option';
import { EpicHeader } from '../../../EpicTable/cells/EpicHeader/EpicHeader';
import { Select } from '../../../../form/Select/Select';
import { ReactNode } from 'react';

type Props<T> = {
  /**
   * The width of the cell in pixels.
   */
  width: number;

  /**
   * The height of the cell in pixels.
   * Defaults to 88 when onSearch is defined, or 44 otherwise.
   */
  height?: number;

  /**
   * Optionally a callback for when the width has changed. By setting
   * this callback you enable the resizing of the EpicHeader.
   *
   * You can never resize the width to less than the original width to
   * prevent columns from becoming too small.
   *
   * When this callback is called you should store the `width` and
   * pass it back into the CrudHeader as the `width` property.
   */
  onResize?: (width: number) => void;

  /**
   * The content of the cell.
   */
  children: ReactNode;

  /**
   * Optionally the current sorting direction.
   * Defaults to NONE.
   */
  direction?: EpicTableSortDirection;

  /**
   * Optional callback which is triggered when the direction changes.
   */
  onSort?: (direction: EpicTableSortDirection) => void;

  /**
   * Optional callback that is triggered when the search value changes.
   * When `options` is defined, a dropdown is displayed, otherwise an open
   * text input is displayed.
   */
  onSearch?: (value: T) => void;

  /**
   * Optionally the current search value.
   */
  searchValue?: T;

  /**
   * Optional search options used when you want to display a dropdown.
   */
  options?: T[];

  /**
   * Optional callback to convert a value of type T to an option to show
   * to the user.
   * Required when `options` is defined.
   */
  labelForOption?: LabelForOption<T>;

  /**
   * Optionally the (invisible) label for the search field.
   */
  searchLabel?: ReactNode;
};

/**
 * CrudHeader is used to display a header in the EpicTable or CrudTable.
 * If onSort is defined, a sorting button is displayed next to the label.
 * If onSearch is defined, a search field is displayed below the label.
 */
export function CrudHeader<T = string>({
  direction = 'NONE',
  onSort,
  onSearch,
  searchValue,
  options,
  labelForOption,
  width,
  height,
  onResize,
  searchLabel,
  children
}: Props<T>) {
  const header = onSort ? (
    <EpicCellLayout mode="horizontal">
      {children}
      <EpicSort direction={direction} onChange={onSort} />
    </EpicCellLayout>
  ) : (
    <>{children}</>
  );

  if (onSearch) {
    return (
      <EpicHeader
        width={width}
        height={height ? height : 88}
        onResize={onResize}
      >
        <EpicCellLayout mode="vertical">
          {header}
          {options && labelForOption ? (
            <Select<T>
              value={searchValue}
              options={options}
              labelForOption={labelForOption}
              onChange={onSearch}
              label={searchLabel}
              hiddenLabel={true}
            />
          ) : (
            <Input
              // @ts-expect-error Default value type is a string
              value={searchValue}
              // @ts-expect-error Default value type is a string
              onChange={onSearch}
              type="search"
              label={searchLabel}
              hiddenLabel={true}
            />
          )}
        </EpicCellLayout>
      </EpicHeader>
    );
  }

  return (
    <EpicHeader width={width} height={height} onResize={onResize}>
      {header}
    </EpicHeader>
  );
}
