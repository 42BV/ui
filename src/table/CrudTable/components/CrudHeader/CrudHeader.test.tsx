import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { CrudHeader } from './CrudHeader';
import * as UseId from '../../../../hooks/useId/useId';
import userEvent from '@testing-library/user-event';

describe('Component: CrudHeader', () => {
  function setup({
    hasHeight,
    hasResize,
    hasSort,
    hasSearch,
    hasSearchOptions
  }: {
    hasHeight?: boolean;
    hasResize?: boolean;
    hasSort?: boolean;
    hasSearch?: boolean;
    hasSearchOptions?: boolean;
  }) {
    const onResizeSpy = jest.fn();
    const onSearchSpy = jest.fn();
    const onSortSpy = jest.fn();

    const props = {
      width: 300,
      height: hasHeight ? 100 : undefined,
      onResize: hasResize ? onResizeSpy : undefined,
      onSearch: hasSearch ? onSearchSpy : undefined,
      onSort: hasSort ? onSortSpy : undefined,
      options: hasSearchOptions ? [ 'crud', 'header' ] : undefined,
      labelForOption: hasSearchOptions ? (option) => option : undefined
    };

    const { container } = render(
      <CrudHeader {...props}>
        This is a header
      </CrudHeader>
    );

    return { container, onResizeSpy, onSearchSpy, onSortSpy };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('with height', () => {
      const { container } = setup({ hasHeight: true });
      expect(container.firstChild).toHaveStyle({ height: '100px' });
    });

    test('with resize', () => {
      setup({ hasResize: true });
      expect(screen.queryByTestId('epic-table-header-resize')).toBeInTheDocument();
    });

    test('with sort', () => {
      setup({ hasSort: true });
      expect(screen.queryByText('sort')).toBeInTheDocument();
    });

    test('with search', () => {
      setup({ hasSearch: true });
      expect(screen.queryByRole('searchbox')).toBeInTheDocument();
    });

    test('with search and height', () => {
      const { container } = setup({ hasSearch: true, hasHeight: true });
      expect(screen.queryByRole('searchbox')).toBeInTheDocument();
      expect(container.firstChild).toHaveStyle({ height: '100px' });
    });

    test('with search options', () => {
      jest.spyOn(UseId, 'useId').mockReturnValue('test');

      setup({ hasSearch: true, hasSearchOptions: true });

      expect(screen.queryByRole('combobox')).toBeInTheDocument();
      expect(screen.queryByRole('option', { name: 'crud' })).toBeInTheDocument();
      expect(screen.queryByRole('option', { name: 'header' })).toBeInTheDocument();
    });
  });

  describe('events', () => {
    it('should call onSort when the user clicks the sorting button', () => {
      const { onSortSpy } = setup({ hasSort: true });

      fireEvent.click(screen.getByText('sort'));

      expect(onSortSpy).toBeCalledTimes(1);
      expect(onSortSpy).toBeCalledWith('ASC');
    });

    it('should call onSearch when the user types in the search field', () => {
      const { onSearchSpy } = setup({ hasSearch: true });

      fireEvent.change(screen.getByRole('searchbox'), { target: { value: 'test' } });

      expect(onSearchSpy).toBeCalledTimes(1);
      expect(onSearchSpy).toBeCalledWith('test');
    });

    it('should call onSearch when the user selects a search option', async () => {
      expect.assertions(2);

      const { onSearchSpy } = setup({ hasSearch: true, hasSearchOptions: true });

      await userEvent.selectOptions(screen.getByRole('combobox'), 'crud');

      expect(onSearchSpy).toBeCalledTimes(1);
      expect(onSearchSpy).toBeCalledWith('crud');
    });
  });
});
