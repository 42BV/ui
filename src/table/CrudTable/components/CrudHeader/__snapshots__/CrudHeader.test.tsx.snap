// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CrudHeader ui default 1`] = `
<div>
  <div
    class="epic-table-header d-flex align-items-center justify-content-between p-1"
    style="min-width: 300px; width: 300px; height: 44px;"
  >
    This is a header
  </div>
</div>
`;
